sharding/xorValidatorShuffler.go

```diff
@@ -96,134 +103,181 @@ func (rxs *randXORShuffler) UpdateNodeLists(args ArgsUpdateNodes) ResUpdateNodes
 		log.Warn("distributeValidators newNodes failed", "error", err)
 	}
 
-	err = distributeValidators(shuffledOutNodes, waiting, randomness)
+	err = distributor.DistributeValidators(newWaiting, shuffledOutMap, randomness)
 	if err != nil {
-		log.Warn("distributeValidators shuffledOutNodes failed", "error", err)
+		log.Warn("distributeValidators newNodes failed", "error", err)
 	}
 
-	actualLeaving, _ := removeValidatorsFromList(leaving, stillRemainingInLeaving, len(stillRemainingInLeaving))
+	actualLeaving, _ := removeValidatorsFromList(allLeaving, stillRemainingInLeaving, len(stillRemainingInLeaving))
 
 	return ResUpdateNodes{
-		Eligible:       eligible,
-		Waiting:        waiting,
+		Eligible:       newEligible,
+		Waiting:        newWaiting,
 		Leaving:        actualLeaving,
 		StillRemaining: stillRemainingInLeaving,
 	}
 }
 
-// shuffleNodesIntraShards shuffles nodes only between the waiting list and eligible list of a shard, not between
-// different shards
-func shuffleNodesIntraShards(
+func removeLeavingNodesNotExistingInEligibleOrWaiting(
+	leavingValidators []Validator,
+	waiting map[uint32][]Validator,
+	eligible map[uint32][]Validator,
+) ([]Validator, []Validator) {
+	notFoundValidators := make([]Validator, 0)
+
+	for _, v := range leavingValidators {
+		if !foundInMap(waiting, v) {
+			notFoundValidators = append(notFoundValidators, v)
+		}
+	}
+
+	for _, v := range leavingValidators {
+		if !foundInMap(eligible, v) {
+			notFoundValidators = append(notFoundValidators, v)
+		}
+	}
+
+	return removeValidatorsFromList(leavingValidators, notFoundValidators, len(notFoundValidators))
+}
+
+func removeLeavingNodesFromValidatorMaps(
 	eligible map[uint32][]Validator,
 	waiting map[uint32][]Validator,
+	numToRemove map[uint32]int,
 	leaving []Validator,
-	newNodes []Validator,
-	randomness []byte,
-) ResUpdateNodes {
+) (map[uint32][]Validator, map[uint32][]Validator, []Validator) {
 
-	shuffledOutMap := make(map[uint32][]Validator)
 	stillRemainingInLeaving := make([]Validator, len(leaving))
 	copy(stillRemainingInLeaving, leaving)
 
-	waiting, stillRemainingInLeaving = removeFromWaiting(waiting, stillRemainingInLeaving)
-
-	sortedShardIds := sortKeys(eligible)
-	for _, shardId := range sortedShardIds {
-		validators := eligible[shardId]
-		shuffledOutMap[shardId], eligible[shardId], stillRemainingInLeaving = shuffleOutShard(
-			validators,
-			len(waiting[shardId]),
-			stillRemainingInLeaving,
-			randomness,
-		)
-	}
+	waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
+	eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+	return eligible, waiting, stillRemainingInLeaving
+}
 
-	moveNodesToMap(eligible, waiting)
-	err := distributeValidators(newNodes, waiting, randomness)
-	if err != nil {
-		log.Warn("distributeValidators newNodes failed", "error", err)
+func foundInMap(nodesMap map[uint32][]Validator, validator Validator) bool {
```

Comment:
duplicated function. there is also a searchInMap.

Reply:
fixed